jetbrains-core/it/software/aws/toolkits/jetbrains/services/lambda/java/JavaTestUtils.kt

```diff
@@ -50,6 +75,68 @@ internal fun HeavyJavaCodeInsightTestFixtureRule.setUpGradleProject(): PsiClass
             }
             """.trimIndent()
     )
+
+    val jdkHome = IdeaTestUtil.requireRealJdkHome()
+    VfsRootAccess.allowRootAccess(fixture.testRootDisposable, jdkHome)
+    val jdkHomeDir = LocalFileSystem.getInstance().refreshAndFindFileByPath(jdkHome)!!
+    val jdkName = "Gradle JDK"
+    val jdk = SdkConfigurationUtil.setupSdk(emptyArray(), jdkHomeDir, JavaSdk.getInstance(), false, null, jdkName)!!
+
+    WriteAction.runAndWait<Nothing> {
+        ProjectJdkTable.getInstance().addJdk(jdk)
+    }
+
+    Disposer.register(
+        fixture.testRootDisposable,
+        Disposable { WriteAction.runAndWait<Nothing> { ProjectJdkTable.getInstance().removeJdk(jdk) } })
+
+    ExternalSystemApiUtil.subscribe(
+        project,
+        GradleConstants.SYSTEM_ID,
+        object : ExternalSystemSettingsListenerAdapter<GradleProjectSettings>() {
+            override fun onProjectsLinked(settings: Collection<GradleProjectSettings>) {
+                super.onProjectsLinked(settings)
+                settings.first().gradleJvm = jdkName
+            }
+        })
+
+    val gradleProjectSettings = GradleProjectSettings().apply {
+        withQualifiedModuleNames()
+        externalProjectPath = buildFile.path
+    }
+
+    val externalSystemSettings = ExternalSystemApiUtil.getSettings(project, GradleConstants.SYSTEM_ID)
+    externalSystemSettings.setLinkedProjectsSettings(setOf(gradleProjectSettings))
+
+    val error = Ref.create<String>()
+
+    val refreshCallback = object : ExternalProjectRefreshCallback {
+        override fun onSuccess(externalProject: DataNode<ProjectData>?) {
+            if (externalProject == null) {
+                System.err.println("Got null External project after import")
+                return
+            }
+            ServiceManager.getService(ProjectDataManager::class.java).importData(externalProject, project, true)
+            println("External project was successfully imported")
+        }
+
+        override fun onFailure(errorMessage: String, errorDetails: String?) {
+            error.set(errorMessage)
+        }
+    }
+
+    val importSpecBuilder = ImportSpecBuilder(project, GradleConstants.SYSTEM_ID)
+        .callback(refreshCallback)
+        .forceWhenUptodate()
+        .use(ProgressExecutionMode.MODAL_SYNC)
+
+    ExternalSystemUtil.refreshProjects(importSpecBuilder)
```

Comment:
does this happen synchronously?

Reply:
Yes, that is controlled by the ProgressExecutionMode enum 2 lines up